### ISSUE TITLE
What is the best way to pass priority to tasks?

### ISSUE THREAD

**Issue by mindojo-victor** [coleifer/huey#274]
As you know I've my own storage for tasks. So in `dequeue` I can fetch the next tasks by `priority` when I add this attribute to tasks.

The question is what is best way to pass this argument to `huey.storage.BaseStorage#enqueue`?

The same, I think, applies to adding any other arguments to storage, for example `queue`, `retries`, `timeout`, etc.

Celery uses `apply_async` for this. What about Huey? Is there an existing similar mechanism?

**Comment by coleifer** [comment]
I'm not familiar with celery, so I can't comment on other approaches.

As of af9ad0ab867e1458892c6869a021f081c99d39aa any kwargs passed to the task() decorator will get passed on to the task class. See 8a98bcad6d8c37e7da2898cce05a09c55e355ee2 for an example.

Does this address your use-case?

**Comment by mindojo-victor** [comment]
Looks like it does. Thank you! Will test it later.

**Comment by coleifer** [comment]
Oh good, any settings will apply to all instances of the given task, so I didn't know if you were trying to get per-instance priority or just on a per-task-class basis.

**Comment by mindojo-victor** [comment]
Yeah, the main use case is per-instance priority.

**Comment by mindojo-victor** [comment]
Unfortunately `task_settings` is not stored anywhere in the db. At least I could not find it. The additional attributes are just made attributes of `huey.api.QueueTask` but they are not stored in the db. So if I add a `timeout` attribute, or `priority` I cannot get this info in consumers.

**Comment by coleifer** [comment]
Right, in `registry` you can see how a task becomes serialized. It basically just dumps:

* task id
* task name
* execution time (if relevant)
* number of retries
* retry delay
* task data (i.e. arguments passed to task)
* on complete (i.e. task to run next if set)

The task settings are not passed along, because they are tied to a task-class, as opposed to an instance (or invocation) of a task.

**Comment by mindojo-victor** [comment]
so the only way to accomplish per task timeouts or priorities is to pass this metadata with the task data under some standard key? Then the consumer would introspect the data and do what it needs? Doesn't look nice mixing task metadata and data, but I'll go this way.

**Comment by coleifer** [comment]
All the information I listed is on a per-task basis. Or, per-invocation, if you prefer. That all gets serialized to form the message that's stuck in the queue -- all the relevant task data *and* metadata, including how many retries are left, etc.

**Comment by mindojo-victor** [comment]
Thanks for the explanation.
